### ISSUE TITLE
rrf/setSubmitFailed

### ISSUE THREAD

**Issue by ffxsam** [davidkpiano/react-redux-form#185]
I'm stuck and having trouble figuring out what I'm doing wrong. Code below:

```
const MUITextField = createFieldClass({
  TextField: props => ({
    onChange: props.onChange,
    value: props.modelValue,
  }),
});

// ....



        attachmentEditArea = <div>
          <MUITextField model="attachmentModel.content">
            <TextField
              ref="textfield"
              hintText="Enter content here"
              multiLine
              rows={6}
              style={styles.textarea.outer}
              textareaStyle={styles.textarea.text}
            />
          </MUITextField>
          <div>
            <RaisedButton type="submit" primary label="Save" />
          </div>
        </div>;

    return (
      <Form model="attachmentModel" onSubmit={this.newAttachment}>
        <div>
          <DropDownMenu
            value={this.state.chosenType}
            onChange={this.chooseType}
          >
            {types}
          </DropDownMenu>
        </div>

        <div>
          {attachmentEditArea}
        </div>
      </Form>
   )
```

`newAttachment()` is never even called. There are no console errors, so I can't really tell what's happening here, other than in Redux I see an `rrf/setSubmitFailed` being dispatched.

**Comment by davidkpiano** [comment]
`SET_SUBMIT_FAILED` occurs on `<Form>` when the form state is invalid. If you're using `redux-logger`, can you check to see what's making the form invalid?

**Comment by ffxsam** [comment]
I've got the Redux Dev Tools Chrome extension which shows all dispatches & state. What in particular should I look for? (there's a lot!)

**Comment by davidkpiano** [comment]
Look for `attachmentModelForm` (or wherever you set up your `formReducer` for `"attachmentModel"`). Find where it changes from `.valid = true` to `.valid = false`

**Comment by ffxsam** [comment]
That's the really odd thing. `attachmentForm.valid` is always true.

**Comment by ffxsam** [comment]
![redux_devtools](https://cloud.githubusercontent.com/assets/12532733/15034822/b5732516-122e-11e6-94dd-9111e51324fe.png)

**Comment by ffxsam** [comment]
The other fishy thing is that `attachmentModel.errors` is `true`, yet if I drill down into the details, the `errors` object is empty.

**Comment by davidkpiano** [comment]
Can you make a minimal code example demonstrating the issue on esnextb.in ?

**Comment by ffxsam** [comment]
Wow, cool site! I'll try to put something together in the morning.. and hope that I can replicate it. Thanks!

**Comment by ffxsam** [comment]
I can't seem to replicate it in a new project.. argh! How can I go about debugging this, any suggestions?

**Comment by davidkpiano** [comment]
In `form-component.js` there is a few lines of code that _should_ be called since you don't have validators:

```js
    if (!validators && onSubmit && formValid) {
      onSubmit(modelValue);

      return modelValue;
    }
```

so the fact that `onSubmit` isn't being called means that `!validators && onSubmit && formValid` is falsey.

If you go into `lib/components/form-component.js` (if you're using Webpack) you can directly `console.log()` in this file to dig around and see what's going on.

Let me know if you're able to do so.

**Comment by ffxsam** [comment]
Ahh, I got it! It's because the `<Field>` (or in this case `<MUITextField>`) doesn't exist when the page first loads up, so `<Form>` doesn't process it. But when the user chooses the attachment type "text", it adds the `<Field>` but it's too late since `<Form>` has already traversed through its children.

**Comment by ffxsam** [comment]
Nope. Bummer. I had some conditional code that would either show the field or not.. but I simplified my render's return statement to this:

```jsx
return <div className="AttachmentCreator">
      <RaisedButton label="Add Attachment" onTouchTap={this.openAdd} />

      <Form
        model="attachmentModel"
        onSubmit={data => console.log(data)}
      >
        <MUITextField model="attachmentModel.content">
          <TextField
            ref="textfield"
            hintText="Enter content here"
            multiLine
            rows={6}
            style={styles.textarea.outer}
            textareaStyle={styles.textarea.text}
          />
        </MUITextField>
        <div>
          <RaisedButton type="submit" primary label="Save" />
        </div>
      </Form>
    </div>
```

And I'm still getting the `setSubmitFailed`. There's nothing fancy going on with the reducers, pretty basic. I'll try your debugging tips and see if I can figure this out.

**Comment by ffxsam** [comment]
@davidkpiano Ok, `formValid` is false. Very strange.

**Comment by ffxsam** [comment]
Ok, I think a co-worker might've done something wrong. `formValue.model` is equal to `"messageModel"` which is not mine!

**Comment by ffxsam** [comment]
Solved! Ok, maybe you can tell me what was happening here.

A co-worker of mine had another model called `messageModel` with a field called "content." The field name is the same as my `attachmentModel.content`, though this part was irrelevant.

However, he had this:

```jsx
      <Form
        model="messageModel"
        validators={validators}
        onSubmit={message => this.sendMessage(message)}
      >
```

Validators:

```javascript
const validators = {
  content: {
    filled(value) {
      return value && value.length > 0;
    },
  },
};
```

This validator was somehow acting globally, and preventing my form from submitting.

**Comment by ffxsam** [comment]
So it seems there's some cross-contamination between forms. `messageModel`'s validators definitely shouldn't be evaluated when I'm submitting `attachmentModel`.

**Comment by davidkpiano** [comment]
Weird! It's hard to diagnose the problem without a way to replicate, and there should be no cross-contamination. I'll try to see what I can find.

**Comment by ffxsam** [comment]
Well, I noticed the guide no longer demonstrates `validators` as a prop on `<Form>`.. so that might be outdated usage. I haven't tried putting validators on the `<Field>` instead. Honestly, since we're still in alpha, I'll probably just wait to add validators until you launch 1.0. We can't keep up with your progress! :)

**Comment by davidkpiano** [comment]
Okay - `validators` continue to work in `<Form>` so I'll help try to figure out what's going on. The unit tests should cover all possible edge cases, but I might have missed something.

**Comment by davidkpiano** [comment]
@ffxsam Were you able to work around this? I am still not able to recreate the issue without an isolated code example.

**Comment by ffxsam** [comment]
You know, it actually hasn't come up again for some reason. I'm not sure if something on your end changed or on mine.. but it's all good now :)

BTW, thanks for all the time & effort you put into react-redux-form! Much appreciated. If you're a Bitcoin user & share an address, I'm happy to send some beer money ;)

**Comment by davidkpiano** [comment]
Awesome! And no worries, I'm happy to do this for free. This project makes my life easier and (hopefully) other dev's lives easier.

**Comment by markoshust** [comment]
I stumbled on this issue debugging the same issue. It turned out the submit handler I was passing through `handleSubmit` was doing an `e.preventDefault()` check that was leftover from pre-existing code. This was causing the SET_SUBMIT_FAILED action to be triggered. Once I removed the `e.preventDefault()` check/call from my final submit handler, all worked 👌  It would be nice if there was a verbose error message that was telling me this, instead I had to grok through the source to debug.
